uber/models.py

```diff
@@ -1625,6 +1817,464 @@ def must_contact(self):
+            been used.
+        uses_count_display (int): A human readable description of uses_count.
+        uses_remaining (int): Remaining number of times this promo code may
+            be used.
+        uses_remaining_display (int): A human readable description of
+            uses_remaining.
+    """
+
+    FIXED_DISCOUNT = 0
+    FIXED_PRICE = 1
+    PERCENT_DISCOUNT = 2
+    DISCOUNT_OPTS = [
+        (FIXED_DISCOUNT, 'Fixed Discount'),
+        (FIXED_PRICE, 'Fixed Price'),
+        (PERCENT_DISCOUNT, 'Percent Discount')]
+
+    AMBIGUOUS_CHARS = {
+        '0': 'OQD',
+        '1': 'IL',
+        '2': 'Z',
+        '5': 'S',
+        '6': 'G',
+        '8': 'B'}
+
+    UNAMBIGUOUS_CHARS = string.digits + string.ascii_uppercase
+    for _, s in AMBIGUOUS_CHARS.items():
+        UNAMBIGUOUS_CHARS = re.sub('[{}]'.format(s), '', UNAMBIGUOUS_CHARS)
+
+    code = Column(UnicodeText)
+    discount = Column(Integer, nullable=True, default=None)
+    discount_type = Column(Choice(DISCOUNT_OPTS), default=FIXED_DISCOUNT)
+    expiration_date = Column(UTCDateTime, default=c.ESCHATON)
+    uses_allowed = Column(Integer, nullable=True, default=None)
+
+    __table_args__ = (
+        Index('uq_promo_code_normalized_code',
+            func.replace(func.replace(func.lower(code), '-', ''), ' ', ''),
+            unique=True),
+        CheckConstraint(func.trim(code) != '',
+            name='ck_promo_code_non_empty_code'))
+
+    _repr_attr_names = ('code',)
+
+    @property
+    def discount_display(self):
+        if not self.discount:
+            return 'Free badge'
+
+        if self.discount_type == self.FIXED_DISCOUNT:
+            return '${} discount'.format(self.discount)
+        elif self.discount_type == self.FIXED_PRICE:
+            return '${} badge'.format(self.discount)
+        else:
+            return '%{} discount'.format(self.discount)
+
+    @hybrid_property
+    def is_expired(self):
+        return self.expiration_date < localized_now()
+
+    @is_expired.expression
+    def is_expired(cls):
+        return cls.expiration_date < localized_now()
+
+    @hybrid_property
+    def is_unlimited(self):
+        return self.uses_allowed is None
+
+    @is_unlimited.expression
+    def is_unlimited(cls):
+        return cls.uses_allowed == None
+
+    @hybrid_property
+    def is_valid(self):
+        return not self.is_expired and (
+            self.is_unlimited or self.uses_remaining > 0)
+
+    @is_valid.expression
+    def is_valid(cls):
+        return (cls.expiration_date >= localized_now()) & (
+            (cls.uses_allowed == None) | (cls.uses_remaining > 0))
+
+    @hybrid_property
+    def normalized_code(self):
+        return self.normalize_code(code)
+
+    @normalized_code.expression
+    def normalized_code(cls):
+        return func.replace(
+            func.replace(func.lower(cls.code), '-', ''), ' ', '')
+
+    @property
+    def uses_allowed_display(self):
```

Comment:
Again, this is a point that I agree with! But again, I don't believe it's applicable in this case.

It's entirely reasonable for objects to know how to create string representations of themselves. That's why we have the `__str__` and `__repr__` methods. It's also why `datetime` objects have their own string formatting methods.

The thing we want to avoid is embedding *view* logic in our models. When I say view logic, I'm talking about an HTML representation, not just a string representation. So it would be bad to have something like:

```python
@property
def discount_display(self):
    return '<b>{}</b>: {}'.format(self.discount_type, self.discount)
```

The *reason* that's bad is because we can't be sure that `discount_display` will always be used in an HTML renderer. If we ever built a native mobile app for uber, we'd either end up with a bunch of HTML tags in our interface, or we'd have to write specific properties for each type of view. Which is bad. But a plain ole string representation will work in both.

As a side note, you see this issue crop up in lots of places. Reddit uses `&nbsp;` to create blank lines in their posts, so you'll see a bunch of raw `&nbsp;` in their mobile app. That's bad.

The `*_display` properties are perhaps poorly named. They may be better off using a suffix like `*_str`. I think I'll make that change 😄